File: frontend/src/types/quotation.ts
/**
 * Back-office quotations.
 *
 * 🔴 Money here is INTEGER MINOR UNITS as a NUMBER (sales-side convention,
 * like orders), not the decimal strings the procurement types use. Prices
 * INCLUDE VAT: `tax_minor` is the tax contained in `total_minor`, not an
 * amount added to it. `formatMoney` takes minor units directly.
 */

export type QuotationStatus =
  | "draft"
  | "sent"
  | "accepted"
  | "declined"
  | "converted";

/** What a human is shown. `expired` is derived from the date, never stored. */
export type QuotationDisplayStatus = QuotationStatus | "expired";

export interface QuotationItem {
  id: string;
  menu_item_id: string | null;
  name: string;
  description: string | null;
  quantity: number;
  unit_price_minor: number;
  line_total_minor: number;
  display_order: number;
}

export interface Quotation {
  id: string;
  quote_number: string;
  status: QuotationStatus;
  display_status: QuotationDisplayStatus;
  location_id: string | null;
  location_name: string | null;
  customer_id: string | null;
  customer_name: string;
  customer_phone: string | null;
  customer_email: string | null;
  customer_address: string | null;
  customer_trn: string | null;
  issue_date: string;
  valid_until: string;
  tax_rate_bps: number;
  subtotal_minor: number;
  discount_minor: number;
  tax_minor: number;
  total_minor: number;
  notes: string | null;
  terms: string | null;
  sent_at: string | null;
  sent_to_email: string | null;
  email_send_count: number;
  last_email_error: string | null;
  decided_at: string | null;
  decline_reason: string | null;
  converted_order_id: string | null;
  converted_at: string | null;
  created_at: string;
  items: QuotationItem[];
}

export interface QuotationLineCreate {
  menu_item_id?: string | null;
  name?: string | null;
  description?: string | null;
  quantity: number;
  unit_price_minor?: number | null;
}

export interface QuotationCreate {
  customer_name: string;
  lines: QuotationLineCreate[];
  location_id?: string | null;
  customer_phone?: string | null;
  customer_email?: string | null;
  customer_address?: string | null;
  customer_trn?: string | null;
  valid_until?: string | null;
  tax_rate_bps?: number | null;
  discount_minor?: number;
  notes?: string | null;
  terms?: string | null;
}

export interface QuotationUpdate {
  customer_name?: string;
  customer_phone?: string | null;
  customer_email?: string | null;
  customer_address?: string | null;
  customer_trn?: string | null;
  location_id?: string | null;
  valid_until?: string | null;
  tax_rate_bps?: number | null;
  discount_minor?: number | null;
  notes?: string | null;
  terms?: string | null;
  lines?: QuotationLineCreate[];
}

export interface QuotationSendRequest {
  to?: string | null;
  message?: string | null;
  skip_email?: boolean;
}

export interface QuotationSendResult {
  quotation: Quotation;
  email_sent: boolean;
  sent_to: string | null;
  error: string | null;
}

export interface QuotationConversion {
  quotation: Quotation;
  order_id: string;
  order_number: string;
}

File: backend/app/schemas/quotation.py
"""Pydantic schemas for back-office quotations.

🔴 Money is INTEGER MINOR UNITS (sales-side convention, matching orders and the
tax invoice). Prices INCLUDE VAT; `tax_minor` is the VAT contained in the
total, not an amount added to it.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

# `expired` is never stored. It is derived from `valid_until` at read time and
# only ever appears on the way out.
QuotationStatus = Literal["draft", "sent", "accepted", "declined", "converted"]
QuotationDisplayStatus = Literal[
    "draft", "sent", "accepted", "declined", "converted", "expired"
]


class QuotationLineCreate(BaseModel):
    """A line is either a menu item, or free text with its own price."""

    menu_item_id: uuid.UUID | None = None
    name: str | None = Field(None, max_length=300)
    description: str | None = None
    quantity: int = Field(..., ge=1)
    unit_price_minor: int | None = Field(
        None, ge=0, description="Defaults to the menu item's current price"
    )

    @model_validator(mode="after")
    def needs_an_identity(self) -> "QuotationLineCreate":
        if self.menu_item_id is None and not (self.name or "").strip():
            raise ValueError(
                "A line that is not a menu item needs a description of its own."
            )
        if self.menu_item_id is None and self.unit_price_minor is None:
            raise ValueError(
                "A line that is not a menu item needs a price of its own."
            )
        return self


class QuotationCreate(BaseModel):
    customer_name: str = Field(..., max_length=200)
    lines: list[QuotationLineCreate] = Field(..., min_length=1)
    location_id: uuid.UUID | None = None
    customer_id: uuid.UUID | None = None
    customer_phone: str | None = Field(None, max_length=50)
    customer_email: str | None = Field(None, max_length=255)
    customer_address: str | None = None
    customer_trn: str | None = Field(None, max_length=50)
    valid_until: date | None = Field(
        None, description="Defaults to 30 days from today"
    )
    tax_rate_bps: int | None = Field(
        None, ge=0, le=10000, description="Defaults to the restaurant's VAT rate"
    )
    discount_minor: int = Field(default=0, ge=0)
    notes: str | None = None
    terms: str | None = None


class QuotationUpdate(BaseModel):
    customer_name: str | None = Field(None, max_length=200)
    customer_phone: str | None = Field(None, max_length=50)
    customer_email: str | None = Field(None, max_length=255)
    customer_address: str | None = None
    customer_trn: str | None = Field(None, max_length=50)
    location_id: uuid.UUID | None = None
    valid_until: date | None = None
    tax_rate_bps: int | None = Field(None, ge=0, le=10000)
    discount_minor: int | None = Field(None, ge=0)
    notes: str | None = None
    terms: str | None = None
    lines: list[QuotationLineCreate] | None = Field(None, min_length=1)


class QuotationItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    menu_item_id: uuid.UUID | None
    name: str
    description: str | None
    quantity: int
    unit_price_minor: int
    line_total_minor: int
    display_order: int


class QuotationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    quote_number: str
    status: QuotationStatus
    # What to show a human: `expired` when the date has passed and nobody has
    # decided yet. Derived, never stored.
    display_status: QuotationDisplayStatus
    location_id: uuid.UUID | None
    location_name: str | None
    customer_id: uuid.UUID | None
    customer_name: str
    customer_phone: str | None
    customer_email: str | None
    customer_address: str | None
    customer_trn: str | None
    issue_date: date
    valid_until: date
    tax_rate_bps: int
    subtotal_minor: int
    discount_minor: int
    tax_minor: int
    total_minor: int
    notes: str | None
    terms: str | None
    sent_at: datetime | None
    sent_to_email: str | None
    email_send_count: int
    last_email_error: str | None
    decided_at: datetime | None
    decline_reason: str | None
    converted_order_id: uuid.UUID | None
    converted_at: datetime | None
    created_at: datetime
    items: list[QuotationItemResponse]


class QuotationSendRequest(BaseModel):
    to: str | None = Field(None, max_length=255)
    message: str | None = Field(None, max_length=2000)
    skip_email: bool = Field(
        default=False,
        description="Mark it sent without emailing, for one handed over in person.",
    )


class QuotationSendResponse(BaseModel):
    quotation: QuotationResponse
    email_sent: bool
    sent_to: str | None
    error: str | None = None


class QuotationDecision(BaseModel):
    accepted: bool
    reason: str | None = Field(
        None, max_length=500, description="Why, when it was declined"
    )


class QuotationConversion(BaseModel):
    quotation: QuotationResponse
    order_id: uuid.UUID
    order_number: str
